### PR TITLE
MemPostings: allocate ListPostings once in PostingsForLabelMatching

### PR DESCRIPTION
**Pull request by colega**
Instead of allocating ListPostings pointers one by one, allocate a slice and take pointers from that. It's faster, and also generates less garbage (NewListPostings is one of the top offenders in number of allocations).

```
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/tsdb
cpu: Apple M3 Pro
                                                                     │     main      │                 lps                  │
                                                                     │    sec/op     │    sec/op     vs base                │
Querier/Head/PostingsForMatchers/n="1"-12                               235.4n ± 12%   225.1n ±  4%   -4.35% (p=0.018 n=10)
Querier/Head/PostingsForMatchers/n="X"-12                               220.8n ±  2%   225.5n ±  1%   +2.11% (p=0.008 n=10)
Querier/Head/PostingsForMatchers/n="1",j="foo"-12                       329.5n ±  2%   329.5n ±  1%        ~ (p=0.897 n=10)
Querier/Head/PostingsForMatchers/n="X",j="foo"-12                       243.6n ±  5%   231.9n ±  0%   -4.80% (p=0.003 n=10)
Querier/Head/PostingsForMatchers/j="foo",n="1"-12                       328.9n ±  6%   329.5n ±  0%        ~ (p=0.617 n=10)
Querier/Head/PostingsForMatchers/n="1",j!="foo"-12                      330.1n ±  2%   327.0n ±  1%   -0.94% (p=0.005 n=10)
Querier/Head/PostingsForMatchers/n="1",i!="2"-12                        348.4n ±  3%   337.2n ±  3%   -3.19% (p=0.016 n=10)
Querier/Head/PostingsForMatchers/n="X",j!="foo"-12                      235.9n ±  7%   235.2n ±  1%        ~ (p=1.000 n=10)
Querier/Head/PostingsForMatchers/i=~"1[0-9]",j=~"foo|bar"-12            639.0n ±  4%   649.1n ±  1%   +1.57% (p=0.020 n=10)
Querier/Head/PostingsForMatchers/j=~"foo|bar"-12                        360.6n ± 16%   357.7n ±  4%        ~ (p=0.072 n=10)
Querier/Head/PostingsForMatchers/j=~"XXX|YYY"-12                        302.5n ±  2%   306.4n ±  2%   +1.31% (p=0.014 n=10)
Querier/Head/PostingsForMatchers/j=~"X.+"-12                            292.4n ±  5%   284.8n ±  0%   -2.63% (p=0.003 n=10)
Querier/Head/PostingsForMatchers/i=~"(1|2|3|4|5|6|20|55)"-12            599.4n ± 15%   573.1n ±  1%   -4.40% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i!~"(1|2|3|4|5|6|20|55)"-12            633.2n ±  8%   604.3n ±  3%   -4.57% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~"X|Y|Z"-12                          358.3n ±  3%   350.1n ±  3%        ~ (p=0.063 n=10)
Querier/Head/PostingsForMatchers/i!~"X|Y|Z"-12                          402.6n ±  3%   385.1n ±  4%   -4.35% (p=0.006 n=10)
Querier/Head/PostingsForMatchers/i=~".*"-12                             193.4n ±  3%   190.0n ±  3%        ~ (p=0.149 n=10)
Querier/Head/PostingsForMatchers/i=~"1.*"-12                            4.260m ±  5%   3.636m ±  2%  -14.64% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~".*1"-12                            3.229m ±  3%   2.901m ±  2%  -10.18% (p=0.001 n=10)
Querier/Head/PostingsForMatchers/i=~".+"-12                            10.821m ±  4%   7.898m ±  3%  -27.02% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~".+",j=~"X.+"-12                   10.658m ±  3%   7.762m ±  3%  -27.17% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~""-12                               6.227m ±  2%   6.039m ±  2%   -3.02% (p=0.003 n=10)
Querier/Head/PostingsForMatchers/i!=""-12                               6.244m ±  3%   5.992m ±  5%   -4.05% (p=0.029 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".*",j="foo"-12               352.6n ±  6%   341.4n ±  2%   -3.18% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="X",i=~".*",j="foo"-12               258.5n ±  8%   245.7n ±  5%   -4.95% (p=0.005 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".*",i!="2",j="foo"-12        476.1n ±  6%   457.9n ± 11%        ~ (p=0.063 n=10)
Querier/Head/PostingsForMatchers/n="1",i!=""-12                         6.519m ±  8%   5.908m ±  3%   -9.38% (p=0.001 n=10)
Querier/Head/PostingsForMatchers/n="1",i!="",j="foo"-12                 6.498m ±  5%   6.055m ±  4%   -6.82% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"X.+"-12                6.475m ±  8%   5.884m ± 10%   -9.14% (p=0.003 n=10)
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"XXX|YYY"-12            6.598m ±  6%   5.927m ±  2%  -10.17% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~"X|Y|Z",j="foo"-12            498.4n ±  5%   467.6n ±  5%   -6.19% (p=0.001 n=10)
Querier/Head/PostingsForMatchers/n="1",i!~"X|Y|Z",j="foo"-12            579.8n ±  5%   537.0n ±  2%   -7.38% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",j="foo"-12              10.801m ± 11%   7.707m ± 12%  -28.65% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~"1.+",j="foo"-12              4.931m ±  8%   3.680m ±  2%  -25.37% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".*1.*",j="foo"-12           10.747m ±  9%   8.958m ±  6%  -16.64% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!="2",j="foo"-12       10.874m ±  5%   7.715m ±  6%  -29.05% (p=0.002 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-12      17.49m ±  4%   12.67m ±  3%  -27.57% (p=0.002 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~".*2.*",j="foo"-12    21.84m ±  2%   19.04m ±  6%  -12.84% (p=0.002 n=10)
Querier/Head/PostingsForMatchers/n="X",i=~".+",i!~".*2.*",j="foo"-12    296.1n ±  0%   296.5n ±  4%        ~ (p=0.210 n=10)
geomean                                                                 21.42µ         19.57µ         -8.62%

                                                                     │     main     │                  lps                  │
                                                                     │     B/op     │     B/op      vs base                 │
Querier/Head/PostingsForMatchers/n="1"-12                                64.00 ± 0%     64.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X"-12                                48.00 ± 0%     48.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",j="foo"-12                        176.0 ± 0%     176.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X",j="foo"-12                        48.00 ± 0%     48.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j="foo",n="1"-12                        176.0 ± 0%     176.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",j!="foo"-12                       176.0 ± 0%     176.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="2"-12                         176.0 ± 0%     176.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X",j!="foo"-12                       48.00 ± 0%     48.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"1[0-9]",j=~"foo|bar"-12             512.0 ± 0%     512.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j=~"foo|bar"-12                         352.0 ± 0%     352.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j=~"XXX|YYY"-12                         128.0 ± 0%     128.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j=~"X.+"-12                             80.00 ± 0%     80.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"(1|2|3|4|5|6|20|55)"-12             416.0 ± 0%     416.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i!~"(1|2|3|4|5|6|20|55)"-12             480.0 ± 0%     480.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"X|Y|Z"-12                           176.0 ± 0%     176.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i!~"X|Y|Z"-12                           240.0 ± 0%     240.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~".*"-12                              64.00 ± 0%     64.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"1.*"-12                           2.722Mi ± 0%   2.727Mi ± 0%  +0.17% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~".*1"-12                           1.531Mi ± 0%   1.531Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~".+"-12                            12.22Mi ± 0%   12.23Mi ± 0%  +0.02% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~".+",j=~"X.+"-12                   12.22Mi ± 0%   12.23Mi ± 0%  +0.02% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~""-12                              10.69Mi ± 0%   10.69Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i!=""-12                              10.69Mi ± 0%   10.69Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~".*",j="foo"-12                176.0 ± 0%     176.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X",i=~".*",j="foo"-12                48.00 ± 0%     48.00 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~".*",i!="2",j="foo"-12         288.0 ± 0%     288.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!=""-12                        10.69Mi ± 0%   10.69Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="",j="foo"-12                10.69Mi ± 0%   10.69Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"X.+"-12               10.69Mi ± 0%   10.69Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"XXX|YYY"-12           10.69Mi ± 0%   10.69Mi ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~"X|Y|Z",j="foo"-12             240.0 ± 0%     240.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!~"X|Y|Z",j="foo"-12             352.0 ± 0%     352.0 ± 0%       ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~".+",j="foo"-12              12.22Mi ± 0%   12.23Mi ± 0%  +0.02% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~"1.+",j="foo"-12             2.722Mi ± 0%   2.727Mi ± 0%  +0.17% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".*1.*",j="foo"-12           5.906Mi ± 0%   5.907Mi ± 0%  +0.00% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!="2",j="foo"-12       12.22Mi ± 0%   12.23Mi ± 0%  +0.02% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-12     14.95Mi ± 0%   14.96Mi ± 0%  +0.05% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~".*2.*",j="foo"-12   18.14Mi ± 0%   18.14Mi ± 0%  +0.02% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="X",i=~".+",i!~".*2.*",j="foo"-12     48.00 ± 0%     48.00 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                                                13.07Ki        13.07Ki       +0.01%
¹ all samples are equal

                                                                     │      main       │                  lps                  │
                                                                     │    allocs/op    │  allocs/op   vs base                  │
Querier/Head/PostingsForMatchers/n="1"-12                                   3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X"-12                                   3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",j="foo"-12                           7.000 ± 0%    7.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X",j="foo"-12                           3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j="foo",n="1"-12                           7.000 ± 0%    7.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",j!="foo"-12                          7.000 ± 0%    7.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="2"-12                            7.000 ± 0%    7.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X",j!="foo"-12                          3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"1[0-9]",j=~"foo|bar"-12                14.00 ± 0%    14.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j=~"foo|bar"-12                            8.000 ± 0%    8.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j=~"XXX|YYY"-12                            6.000 ± 0%    6.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/j=~"X.+"-12                                4.000 ± 0%    4.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"(1|2|3|4|5|6|20|55)"-12                12.00 ± 0%    12.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i!~"(1|2|3|4|5|6|20|55)"-12                15.00 ± 0%    15.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"X|Y|Z"-12                              7.000 ± 0%    7.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i!~"X|Y|Z"-12                              10.00 ± 0%    10.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~".*"-12                                 3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~"1.*"-12                            11118.000 ± 0%    8.000 ± 0%  -99.93% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~".*1"-12                                4.000 ± 0%    4.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i=~".+"-12                            100007.000 ± 0%    8.000 ± 0%  -99.99% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~".+",j=~"X.+"-12                    100011.00 ± 0%    12.00 ± 0%  -99.99% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/i=~""-12                                  100.0k ± 0%   100.0k ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/i!=""-12                                  100.0k ± 0%   100.0k ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~".*",j="foo"-12                   7.000 ± 0%    7.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="X",i=~".*",j="foo"-12                   3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~".*",i!="2",j="foo"-12            11.00 ± 0%    11.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!=""-12                            100.0k ± 0%   100.0k ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="",j="foo"-12                    100.0k ± 0%   100.0k ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"X.+"-12                   100.0k ± 0%   100.0k ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!="",j=~"XXX|YYY"-12               100.0k ± 0%   100.0k ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~"X|Y|Z",j="foo"-12                10.00 ± 0%    10.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i!~"X|Y|Z",j="foo"-12                14.00 ± 0%    14.00 ± 0%        ~ (p=1.000 n=10) ¹
Querier/Head/PostingsForMatchers/n="1",i=~".+",j="foo"-12               100014.00 ± 0%    15.00 ± 0%  -99.99% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~"1.+",j="foo"-12               11125.00 ± 0%    15.00 ± 0%  -99.87% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".*1.*",j="foo"-12             40965.00 ± 0%    15.00 ± 0%  -99.96% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!="2",j="foo"-12        100019.00 ± 0%    20.00 ± 0%  -99.98% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-12      111201.00 ± 0%    92.00 ± 0%  -99.92% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="1",i=~".+",i!~".*2.*",j="foo"-12    141044.00 ± 0%    95.00 ± 0%  -99.93% (p=0.000 n=10)
Querier/Head/PostingsForMatchers/n="X",i=~".+",i!~".*2.*",j="foo"-12        3.000 ± 0%    3.000 ± 0%        ~ (p=1.000 n=10) ¹
geomean                                                                     225.7         35.67       -84.19%
¹ all samples are equal
```

### CODE REVIEW

**Comment by colega** [comment]
(Force-pushed to re-trigger CI as something unrelated failed on Windows tests)

File: tsdb/index/postings.go [hunk 1]
```diff
@@ -430,15 +430,17 @@ func (p *MemPostings) PostingsForLabelMatching(ctx context.Context, name string,
...
-	for _, v := range vals {
+	for i, v := range vals {
 		if refs, ok := e[v]; ok {
 			// Some of the values may have been garbage-collected in the meantime this is fine, we'll just skip them.
 			// If we didn't let the mutex go, we'd have these postings here, but they would be pointing nowhere
 			// because there would be a `MemPostings.Delete()` call waiting for the lock to delete these labels,
 			// because the series were deleted already.
-			its = append(its, NewListPostings(refs))
+			lps[i] = ListPostings{list: refs}
+			its = append(its, &lps[i])
```
Comment (colega) [inline tsdb/index/postings.go:443]: This triggers my OCD as well, I originally changed this to `its[i] = &lps[i]`, but then we have nil iterators on some positions.

**Comment by colega** [comment]
Windows tests are also failing on some `main` builds.

**Review by jesusvazquez — APPROVED** [review]
LGTM

**Pull request merged**